### ISSUE TITLE
Indexing a tensor with non-integer number causes silent rounding

### ISSUE THREAD

**Issue by timharley** [torch/torch7#575]
```
t = torch.Tensor({1,2})
print(t[1], t[1.999], t[2])
```

This can cause silent indexing bugs if a calculation doesn't return an exact integer due to floating point precision issues. I propose that indexing a Tensor with a non-integer value causes a Lua error to be raised, rather than silently rounding down via a simple cast. What are peoples thoughts on this?

In terms of implementation, this would involve replacing all or almost all calls to `luaL_checklong` to a new function which performs this checking. e.g. [replace this call](https://github.com/torch/torch7/blob/master/generic/Tensor.c#L884).

**Comment by timharley** [comment]
If this gets a consensus of agreement, we would need to do the same for cutorch too.

**Comment by andreaskoepf** [comment]
It should be noted that 'rounding' for luaL_checklong() means integer truncation, e.g. as in a cast from double to long.
I am not 100% sure but I think this is rather safe. Can you come up with an example in which an integer-index calculation (which immediately drops the fractional part) would have resulted in a higher index than with double-float arithmetic and cast immediately before the index-operation?

**Comment by timharley** [comment]
Yes, my language was not particularly precise. `double` (i.e from a [`lua_Number`](http://pgl.yoyo.org/luai/i/lua_Number)) to `long` truncation is safe, provided the truncated value fits inside the range of a `long`. This issue is not trying to suggest that this truncation is itself a bug, it is more that being able to index a tensor with a non-integer valued `lua_Number` is not intuitive, and it is unclear whether torch should support it at all. If this occurs, it is more likely to be a user error than intentional, and the fact that torch silently truncates means that bugs in user code can be silently accepted. If instead this raised a Lua error, it would highlight the bug in user code or force the user to choose the rounding or truncation behaviour that they desire at the Lua level.

**Comment by andreaskoepf** [comment]
@temerick I see your concerns - you would prefer that a non-integer index value should result in an error instead of automatic-truncation because it reminds the coder to be very precise with his/her requests.

I personally think this is a philosophical question: E.g. on the one hand of the spectrum a system might expect the developer/user to completely obey its rules and on the other hand a system could try to make the best out of the potentially erroneous user input. For example if you take the language R ... adding a character string to a numeric vector converts all the elements in the vector to character. The same is true for lua where you can concat a number and a string with the '..' operator. Maybe implicit conversion between number and string is even a bit more questionable but it is sort of the same 'problem/feature' (depending on your view).
IMO there are two points which support the truncation: 1. It saves the developer from writing a lot of math.floor() (or torch.trunc() which behaves like a cast to long for negative values which would be provided by a proposed [PR](https://github.com/torch/torch7/pull/568)) and 2. it makes using torch as a REPL data analysis language more comfortable (which might also be the reason behind potentially user-friendly decisions that were made for R).

**Comment by soumith** [comment]
I'm okay with it. cc: @andresy
